File: Assets/Bundles/Client/Resources/TS/Controllers/Global/BlockInteractions/BlockHealthController.lua
-- Compiled with unity-ts v2.1.0-75
local Reflect = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Reflect
local Controller = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Controller
local ClientSignals = require("Client/TS/ClientSignals").ClientSignals
local Game = require("Shared/TS/Game").Game
local Network = require("Shared/TS/Network").Network
local _Timer = require("Shared/TS/Util/Timer")
local OnLateUpdate = _Timer.OnLateUpdate
local SetInterval = _Timer.SetInterval
local VoxelDataAPI = require("Shared/TS/VoxelWorld/VoxelData/VoxelDataAPI").VoxelDataAPI
local WorldAPI = require("Shared/TS/VoxelWorld/WorldAPI").WorldAPI
local EffectsManager = require("Shared/TS/Effects/EffectsManager").EffectsManager
local ProgressBarGraphics = require("Shared/TS/UI/ProgressBarGraphics").ProgressBarGraphics
local _ReferenceManagerResources = require("Shared/TS/Util/ReferenceManagerResources")
local BundleGroupNames = _ReferenceManagerResources.BundleGroupNames
local Bundle_Blocks = _ReferenceManagerResources.Bundle_Blocks
local Bundle_Blocks_UI = _ReferenceManagerResources.Bundle_Blocks_UI
local Bundle_Blocks_VFX = _ReferenceManagerResources.Bundle_Blocks_VFX
local Theme = require("Shared/TS/Util/Theme").Theme
local CameraReferences = require("Client/TS/Controllers/Global/Camera/CameraReferences").CameraReferences
local BeforeBlockHitSignal = require("Client/TS/Controllers/Global/BlockInteractions/Signal/BeforeBlockHitSignal").BeforeBlockHitSignal
local BlockHealthController
do
	BlockHealthController = setmetatable({}, {
		__tostring = function()
			return "BlockHealthController"
		end,
	})
	BlockHealthController.__index = BlockHealthController
	function BlockHealthController.new(...)
		local self = setmetatable({}, BlockHealthController)
		return self:constructor(...) or self
	end
	function BlockHealthController:constructor(invController, blockSelectController, entityController)
		self.invController = invController
		self.blockSelectController = blockSelectController
		self.entityController = entityController
		self.blockHealthBars = {}
		self.HEALTHBAR_EXPIRE_TIME = 1.25
	end
	function BlockHealthController:OnStart()
		Network.ServerToClient.BlockHit.Client:OnServerEvent(function(blockPos, entityId)
			if Game.LocalPlayer.Character and entityId == Game.LocalPlayer.Character.id then
				return nil
			end
			local voxel = WorldAPI:GetMainWorld():GetRawVoxelDataAt(blockPos)
			if voxel ~= 0 and (voxel == voxel and voxel) then
				local entity = self.entityController:GetEntityById(entityId)
				local blockId = VoxelWorld:VoxelDataToBlockId(voxel)
				ClientSignals.AfterBlockHit:Fire({
					pos = blockPos,
					blockId = blockId,
					entity = entity,
				})
				self:VisualizeBlockHealth(blockPos)
			end
		end)
		Network.ServerToClient.BlockDestroyed.Client:OnServerEvent(function(blockPos, blockId)
			self:VisualizeBlockBreak(blockPos, blockId)
		end)
		OnLateUpdate:Connect(function(dt)
			local _blockHealthBars = self.blockHealthBars
			local _arg0 = function(healthbarEntry, block)
				healthbarEntry.gameObject.transform.rotation = CameraReferences:Instance().mainCamera.transform.rotation
			end
			for _k, _v in _blockHealthBars do
				_arg0(_v, _k, _blockHealthBars)
			end
		end)
		-- Cleanup health bars after no changes are made
		SetInterval(0.1, function()
			local toRemove = {}
			local _blockHealthBars = self.blockHealthBars
			local _arg0 = function(entry, pos)
				if Time.time >= entry.lastHitTime + self.HEALTHBAR_EXPIRE_TIME then
					local _pos = pos
					local _entry = entry
					toRemove[_pos] = _entry
				end
			end
			for _k, _v in _blockHealthBars do
				_arg0(_v, _k, _blockHealthBars)
			end
			local _arg0_1 = function(entry, pos)
				self:DeleteHealthBar(entry, pos)
			end
			for _k, _v in toRemove do
				_arg0_1(_v, _k, toRemove)
			end
		end)
	end
	function BlockHealthController:OnBeforeBlockHit(voxelPos, block)
		ClientSignals.BeforeBlockHit:Fire(BeforeBlockHitSignal.new(voxelPos, block))
	end
	function BlockHealthController:VisualizeBlockHealth(blockPos)
		-- Get or create health bar
		local _blockHealthBars = self.blockHealthBars
		local _blockPos = blockPos
		local healthBarEntry = _blockHealthBars[_blockPos]
		if not healthBarEntry then
			healthBarEntry = self:AddHealthBar(blockPos)
			if not healthBarEntry then
				return nil
			end
		end
		-- Update the health bars value
		local currentHealth = self:GetBlockHealth(blockPos)
		healthBarEntry.lastHitTime = Time.time
		healthBarEntry.progressBar:SetValue(currentHealth / healthBarEntry.maxHealth)
		if currentHealth > 0 then
			local effect = EffectsManager:SpawnBundleEffect(BundleGroupNames.Blocks, Bundle_Blocks.VFX, Bundle_Blocks_VFX.OnHit, blockPos, Vector3.zero)
			if effect then
				local block = WorldAPI:GetMainWorld():GetBlockAt(blockPos)
				local blockColor = block:GetAverageColor()
				print("block color: " .. tostring(blockColor));
				(effect.transform:GetChild(0):GetComponent("ParticleSystem")).startColor = blockColor
			end
		end
	end
	function BlockHealthController:VisualizeBlockBreak(blockPos, blockId)
		-- Get or create health bar
		local _blockHealthBars = self.blockHealthBars
		local _blockPos = blockPos
		local entry = _blockHealthBars[_blockPos]
		if not entry then
			entry = self:AddHealthBar(blockPos)
		end
		-- Play destruction vfx
		local effect = EffectsManager:SpawnBundleEffect(BundleGroupNames.Blocks, Bundle_Blocks.VFX, Bundle_Blocks_VFX.OnDeath, blockPos, Vector3.zero)
		if effect then
			local blockColor = WorldAPI:GetMainWorld():GetBlockAverageColor(blockId)
			if not blockColor then
				return nil
			end
			(effect.transform:GetChild(0):GetComponent("ParticleSystem")).startColor = blockColor
		end
		-- Make sure the progress bar is at 0
		local _result = entry
		if _result ~= nil then
			_result = _result.progressBar
			if _result ~= nil then
				_result:SetValue(0)
			end
		end
	end
	function BlockHealthController:GetBlockHealth(blockPos)
		local _condition = VoxelDataAPI:GetVoxelData(blockPos, "health")
		if _condition == nil then
			_condition = WorldAPI.DefaultVoxelHealth
		end
		return _condition
	end
	function BlockHealthController:AddHealthBar(blockPos)
		-- Spawn the health bar
		local _fn = EffectsManager
		local _exp = BundleGroupNames.Blocks
		local _exp_1 = Bundle_Blocks.UI
		local _exp_2 = Bundle_Blocks_UI.HealthBar
		local _blockPos = blockPos
		local _vector3 = Vector3.new(0.5, 1.5, 0.5)
		local healthBarGo = _fn:SpawnBundleEffect(_exp, _exp_1, _exp_2, _blockPos + _vector3, Vector3.zero, -1)
		if not healthBarGo then
			error("Missing health bar graphic!")
			return nil
		end
		-- Get the meta for the hit block
		local _itemMeta = WorldAPI:GetMainWorld():GetBlockAt(blockPos)
		if _itemMeta ~= nil then
			_itemMeta = _itemMeta.itemMeta
		end
		local itemMeta = _itemMeta
		-- Create health bar entry
		local _result = itemMeta
		if _result ~= nil then
			_result = _result.block
			if _result ~= nil then
				_result = _result.health
			end
		end
		local _condition = _result
		if _condition == nil then
			_condition = WorldAPI.DefaultVoxelHealth
		end
		local maxHealth = _condition
		local healthBarEntry = {
			gameObject = healthBarGo,
			lastHitTime = Time.time,
			progressBar = ProgressBarGraphics.new(healthBarGo.transform:GetChild(0), {
				initialPercentDelta = self:GetBlockHealth(blockPos) / maxHealth,
				fillColor = Theme.Green,
			}),
			maxHealth = maxHealth,
		}
		local _blockHealthBars = self.blockHealthBars
		local _blockPos_1 = blockPos
		local _healthBarEntry = healthBarEntry
		_blockHealthBars[_blockPos_1] = _healthBarEntry
		return healthBarEntry
	end
	function BlockHealthController:RemoveHealthBar(blockPos)
		local _blockHealthBars = self.blockHealthBars
		local _blockPos = blockPos
		local entry = _blockHealthBars[_blockPos]
		if entry then
			self:DeleteHealthBar(entry, blockPos)
		end
	end
	function BlockHealthController:DeleteHealthBar(entry, blockPos)
		local _blockHealthBars = self.blockHealthBars
		local _blockPos = blockPos
		_blockHealthBars[_blockPos] = nil
		entry.progressBar:OnDelete()
	end
end
-- (Flamework) BlockHealthController metadata
Reflect.defineMetadata(BlockHealthController, "identifier", "Bundles/Client/Controllers/Global/BlockInteractions/BlockHealthController@BlockHealthController")
Reflect.defineMetadata(BlockHealthController, "flamework:parameters", { "Bundles/Client/Controllers/Global/Inventory/InventoryController@InventoryController", "Bundles/Client/Controllers/Global/BlockInteractions/BlockSelectController@BlockSelectController", "Bundles/Client/Controllers/Global/Entity/EntityController@EntityController" })
Reflect.defineMetadata(BlockHealthController, "flamework:implements", { "$:flamework@OnStart" })
Reflect.decorate(BlockHealthController, "$:flamework@Controller", Controller, { {} })
return {
	BlockHealthController = BlockHealthController,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Bundles/Imports/Core/Core~/src/Client/Controllers/BlockInteractions/BlockHealthController.ts
import { Controller, OnStart } from "@easy-games/flamework-core";
import { CoreClientSignals } from "Client/CoreClientSignals";
import { CoreNetwork } from "Shared/CoreNetwork";
import { EffectsManager } from "Shared/Effects/EffectsManager";
import { Game } from "Shared/Game";
import { GameObjectUtil } from "Shared/GameObject/GameObjectUtil";
import { ProgressBarGraphics } from "Shared/UI/ProgressBarGraphics";
import {
	BundleGroupNames,
	Bundle_Blocks,
	Bundle_Blocks_UI,
	Bundle_Blocks_VFX,
} from "Shared/Util/ReferenceManagerResources";
import { Theme } from "Shared/Util/Theme";
import { SetInterval } from "Shared/Util/Timer";
import { BlockDataAPI } from "Shared/VoxelWorld/BlockData/BlockDataAPI";
import { WorldAPI } from "Shared/VoxelWorld/WorldAPI";
import { EntityController } from "../Entity/EntityController";
import { InventoryController } from "../Inventory/InventoryController";
import { BlockSelectController } from "./BlockSelectController";
import { Entity } from "Shared/Entity/Entity";

interface HealthBarEntry {
	gameObject: GameObject;
	progressBar: ProgressBarGraphics;
	lastHitTime: number;
	maxHealth: number;
}

@Controller({})
export class BlockHealthController implements OnStart {
	private blockHealthBars = new Map<Vector3, HealthBarEntry>();
	HEALTHBAR_EXPIRE_TIME = 1.25;

	constructor(
		private readonly invController: InventoryController,
		private readonly blockSelectController: BlockSelectController,
		private readonly entityController: EntityController,
	) {}

	OnStart(): void {
		CoreNetwork.ServerToClient.BlockHit.Client.OnServerEvent((blockPos, blockId, entityId) => {
			if (Game.LocalPlayer.Character && entityId === Game.LocalPlayer.Character.id) return;

			const entity = this.entityController.GetEntityById(entityId);
			if (entity) {
				this.OnBlockHit(entity, blockPos, blockId, true);
			}
		});

		CoreNetwork.ServerToClient.BlockGroupHit.Client.OnServerEvent((blockPositions, blockIds, entityId) => {
			if (Game.LocalPlayer.Character && entityId === Game.LocalPlayer.Character.id) return;

			const entity = this.entityController.GetEntityById(entityId);
			if (entity) {
				blockPositions.forEach((position, index) => {
					this.OnBlockHit(entity, position, blockIds[index], true);
				});
			}
		});

		CoreNetwork.ServerToClient.BlockDestroyed.Client.OnServerEvent((blockPos, blockId) => {
			this.VisualizeBlockBreak(blockPos, blockId);
		});

		CoreNetwork.ServerToClient.BlockGroupDestroyed.Client.OnServerEvent((blockPositions, blockIds) => {
			blockPositions.forEach((position, index) => {
				this.VisualizeBlockBreak(position, blockIds[index], false);
			});
		});

		// OnLateUpdate.Connect((dt) => {
		// 	this.blockHealthBars.forEach((healthbarEntry, block) => {
		// 		healthbarEntry.gameObject.transform.rotation =
		// 			CameraReferences.Instance().mainCamera.transform.rotation;
		// 	});
		// });

		//Cleanup health bars after no changes are made
		SetInterval(0.1, () => {
			const toRemove = new Map<Vector3, HealthBarEntry>();
			this.blockHealthBars.forEach((entry, pos) => {
				if (Time.time >= entry.lastHitTime + this.HEALTHBAR_EXPIRE_TIME) {
					toRemove.set(pos, entry);
				}
			});
			toRemove.forEach((entry, pos) => {
				this.DeleteHealthBar(entry, pos);
			});
		});
	}

	private OnBlockHit(entity: Entity, blockPos: Vector3, blockId: number, isGroupEvent: boolean) {
		CoreClientSignals.AfterBlockHit.Fire({
			blockPos: blockPos,
			blockId,
			entity,
			isGroupEvent,
		});
		this.VisualizeBlockHealth(blockPos, !isGroupEvent);
	}

	public VisualizeBlockHealth(blockPos: Vector3, showHealthbar = true) {
		let currentHealth = this.GetBlockHealth(blockPos);

		//Get or create health bar
		if (showHealthbar) {
			let healthBarEntry = this.blockHealthBars.get(blockPos);
			if (!healthBarEntry) {
				healthBarEntry = this.AddHealthBar(blockPos);
				if (!healthBarEntry) {
					return;
				}
				healthBarEntry.lastHitTime = Time.time;
				healthBarEntry.progressBar.SetValue(currentHealth / healthBarEntry.maxHealth);
			}
		}

		//Update the health bars value
		if (currentHealth > 0) {
			const effect = EffectsManager.SpawnBundleEffect(
				BundleGroupNames.Blocks,
				Bundle_Blocks.VFX,
				Bundle_Blocks_VFX.OnHit,
				blockPos,
				Vector3.zero,
			);
			if (effect) {
				const block = WorldAPI.GetMainWorld()?.GetBlockAt(blockPos);
				if (block) {
					this.ApplyBlockMaterial(block.blockId, effect);
				}
			}
		}
	}

	public VisualizeBlockBreak(blockPos: Vector3, blockId: number, showHealthbars = true): void {
		//Get or create health bar
		let entry = this.blockHealthBars.get(blockPos);
		if (!entry) {
			if (showHealthbars) {
				entry = this.AddHealthBar(blockPos);
			}
		}

		//Play destruction vfx
		const effect = EffectsManager.SpawnBundleEffect(
			BundleGroupNames.Blocks,
			Bundle_Blocks.VFX,
			Bundle_Blocks_VFX.OnDeath,
			blockPos,
			Vector3.zero,
		);
		if (effect) {
			//const blockColor = WorldAPI.GetMainWorld().GetBlockAverageColor(blockId);
			//if (!blockColor) return;
			this.ApplyBlockMaterial(blockId, effect);
		}

		//Make sure the progress bar is at 0
		entry?.progressBar?.SetValue(0);
	}

	private ApplyBlockMaterial(blockId: number, effect: GameObject) {
		let particles = effect.transform.GetChild(0).GetComponent<ParticleSystemRenderer>();
		const world = WorldAPI.GetMainWorld();
		if (!world) return;
		const blockGO = MeshProcessor.ProduceSingleBlock(blockId, world.voxelWorld);
		if (blockGO) {
			const blockRen = blockGO.GetComponent<Renderer>();
			const blockFilter = blockGO.GetComponent<MeshFilter>();
			particles.mesh = blockFilter.mesh;
			particles.sharedMaterial = blockRen.sharedMaterial;
			GameObjectUtil.Destroy(blockGO);
		}
	}

	private GetBlockHealth(blockPos: Vector3) {
		return BlockDataAPI.GetBlockData<number>(blockPos, "health") ?? WorldAPI.DefaultVoxelHealth;
	}

	private AddHealthBar(blockPos: Vector3): HealthBarEntry | undefined {
		//Spawn the health bar
		const healthBarGo = EffectsManager.SpawnBundleEffect(
			BundleGroupNames.Blocks,
			Bundle_Blocks.UI,
			Bundle_Blocks_UI.HealthBar,
			blockPos.add(new Vector3(0.5, 1.5, 0.5)),
			Vector3.zero,
			-1,
		);
		if (!healthBarGo) {
			error("Missing health bar graphic!");
			return undefined;
		}

		//Get the meta for the hit block
		const itemMeta = WorldAPI.GetMainWorld()?.GetBlockAt(blockPos)?.itemMeta;

		//Create health bar entry
		let maxHealth = itemMeta?.block?.health ?? WorldAPI.DefaultVoxelHealth;
		let healthBarEntry = {
			gameObject: healthBarGo,
			lastHitTime: Time.time,
			progressBar: new ProgressBarGraphics(healthBarGo.transform.GetChild(0), {
				initialPercentDelta: this.GetBlockHealth(blockPos) / maxHealth,
				fillColor: Theme.Green,
			}),
			maxHealth: maxHealth,
		};

		this.blockHealthBars.set(blockPos, healthBarEntry);
		return healthBarEntry;
	}

	private RemoveHealthBar(blockPos: Vector3) {
		const entry = this.blockHealthBars.get(blockPos);
		if (entry) {
			this.DeleteHealthBar(entry, blockPos);
		}
	}

	private DeleteHealthBar(entry: HealthBarEntry, blockPos: Vector3) {
		this.blockHealthBars.delete(blockPos);
		entry.progressBar.OnDelete();
	}
}
